### PR TITLE
Option to disable the thread pool/executor service for Circuitbreaker

### PR DESCRIPTION
**Pull request by kkotamar**
This issue is to fix https://github.com/spring-cloud/spring-cloud-circuitbreaker/issues/147 
I am trying to merge this 2.1.x branch since we are using 2.1.x version.

### CODE REVIEW

**Review by ryanjbaxter — CHANGES_REQUESTED** [review]
I think I would rather take the approach of the `Resilience4JCircuitBreakerFactory` creating a circuit breaker with or without an executor service.  Without the executor service it will run it in the current thread.

**Comment by kkotamar** [comment]
I did this since it is only last step of execution that differs either execute in seperate pool or not. Creating another implementation will also have exact same code except last step.

**Comment by ryanjbaxter** [comment]
I agree it will work the current way, but I would rather use the approach of the presence of the executor service deciding the behavior

**Comment by kkotamar** [comment]
I made changes. Please review.

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java [hunk 1]
```diff
@@ -126,10 +130,17 @@ private Resilience4JCircuitBreaker create(String id, String groupName,
 			ExecutorService circuitBreakerExecutorService) {
 		Resilience4JConfigBuilder.Resilience4JCircuitBreakerConfiguration config = getConfigurations()
 				.computeIfAbsent(id, defaultConfiguration);
-		return new Resilience4JCircuitBreaker(id, groupName, config.getCircuitBreakerConfig(),
-				config.getTimeLimiterConfig(), circuitBreakerRegistry, timeLimiterRegistry,
-				circuitBreakerExecutorService, Optional.ofNullable(circuitBreakerCustomizers.get(id)),
-				bulkheadProvider);
+		if (resilience4JConfigurationProperties.isDisableThreadPool()) {
+			return new Resilience4JCircuitBreaker(id, groupName, config.getCircuitBreakerConfig(),
```
Comment (ryanjbaxter) [inline spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java:134]: I would add another constructor that doesn't take an executor service

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java [hunk 2]
```diff
@@ -61,13 +63,15 @@ public Resilience4JCircuitBreakerFactory() {
 	}
 
 	public Resilience4JCircuitBreakerFactory(CircuitBreakerRegistry circuitBreakerRegistry,
-			TimeLimiterRegistry timeLimiterRegistry, Resilience4jBulkheadProvider bulkheadProvider) {
+			TimeLimiterRegistry timeLimiterRegistry, Resilience4jBulkheadProvider bulkheadProvider,
```
Comment (ryanjbaxter) [inline spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java:66]: Add a new constructor

**Review by ryanjbaxter — CHANGES_REQUESTED** [review]

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java (hunk 1 above)
Comment (kkotamar) [reply spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java:134]: Created a new constructor.

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java (hunk 2 above)
Comment (kkotamar) [reply spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java:66]: Can you clarify why does it need a new constructor? 
This class is in only instantiated from Resilience4JAutoConfiguration.java.

**Comment by kkotamar** [comment]
@ryanjbaxter  Responded to comments. Added constructor for Resilience4JCircuitBreaker with out executor service.

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java (hunk 2 above)
Comment (ryanjbaxter) [reply spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java:66]: It is a public class so anyone can extend it, we try not to break people in minor releases

**Comment by kkotamar** [comment]
@ryanjbaxter 
If this PR looks, can you merge and release?

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java [hunk 3]
```diff
@@ -102,25 +102,49 @@ public Resilience4JCircuitBreaker(String id, String groupName,
 		this.bulkheadProvider = bulkheadProvider;
 	}
 
+	public Resilience4JCircuitBreaker(String id, String groupName,
+			io.github.resilience4j.circuitbreaker.CircuitBreakerConfig circuitBreakerConfig,
+			TimeLimiterConfig timeLimiterConfig, CircuitBreakerRegistry circuitBreakerRegistry,
+			TimeLimiterRegistry timeLimiterRegistry,
+			Optional<Customizer<io.github.resilience4j.circuitbreaker.CircuitBreaker>> circuitBreakerCustomizer,
+			Resilience4jBulkheadProvider bulkheadProvider) {
```
Comment (ryanjbaxter) [inline spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java:110]: You can just call `this(...)` passing `null` for the executor service

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java [hunk 4]
```diff
@@ -68,6 +70,19 @@ public Resilience4JCircuitBreakerFactory(CircuitBreakerRegistry circuitBreakerRe
 		this.defaultConfiguration = id -> new Resilience4JConfigBuilder(id)
 				.circuitBreakerConfig(this.circuitBreakerRegistry.getDefaultConfig())
 				.timeLimiterConfig(this.timeLimiterRegistry.getDefaultConfig()).build();
+		this.resilience4JConfigurationProperties = null;
```
Comment (ryanjbaxter) [inline spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java:73]: Call the new constructor passing `new Resilience4JConfigurationProperties()`

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java [hunk 5]
```diff
@@ -126,10 +141,17 @@ private Resilience4JCircuitBreaker create(String id, String groupName,
 			ExecutorService circuitBreakerExecutorService) {
 		Resilience4JConfigBuilder.Resilience4JCircuitBreakerConfiguration config = getConfigurations()
 				.computeIfAbsent(id, defaultConfiguration);
-		return new Resilience4JCircuitBreaker(id, groupName, config.getCircuitBreakerConfig(),
-				config.getTimeLimiterConfig(), circuitBreakerRegistry, timeLimiterRegistry,
-				circuitBreakerExecutorService, Optional.ofNullable(circuitBreakerCustomizers.get(id)),
-				bulkheadProvider);
+		if (resilience4JConfigurationProperties != null && resilience4JConfigurationProperties.isDisableThreadPool()) {
```
Comment (ryanjbaxter) [inline spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java:144]: No need for the null check if you make the change above

**Review by ryanjbaxter — CHANGES_REQUESTED** [review]
Please add some documentation about the new property as well

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java (hunk 3 above)
Comment (kkotamar) [reply spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java:110]: Implemented as suggested.

**Comment by kkotamar** [comment]
@ryanjbaxter  Addressed comments.

**Comment by kkotamar** [comment]
Docs are added.

**Review by ryanjbaxter — APPROVED** [review]

**Pull request merged**